shell/platform/android/io/flutter/embedding/engine/FlutterEngineAndroidLifecycle.java

```diff
@@ -0,0 +1,126 @@
+package io.flutter.embedding.engine;
+
+import android.arch.lifecycle.DefaultLifecycleObserver;
+import android.arch.lifecycle.Lifecycle;
+import android.arch.lifecycle.LifecycleObserver;
+import android.arch.lifecycle.LifecycleOwner;
+import android.arch.lifecycle.LifecycleRegistry;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+/**
+ * Android {@link Lifecycle} that is owned by a {@link FlutterEngine}.
+ * <p>
+ * {@code FlutterEngineAndroidLifecycle} exists so that {@code FlutterPlugin}s can monitor Android
+ * lifecycle events. When the associated {@link FlutterEngine} is running in an {@code Activity},
+ * that {@code Activity}'s {@link Lifecycle} can be set as the {@code backingLifecycle} of this
+ * class, allowing all Flutter plugins to receive the {@code Activity}'s lifecycle events. Likewise,
+ * when the associated {@link FlutterEngine} is running in a {@code Service}, that {@code Service}'s
+ * {@link Lifecycle} can be set as the {@code backingLifecycle}.
+ * <p>
+ * Sometimes a {@link FlutterEngine} exists in a non-lifecycle location, e.g., an {@code Application},
+ * {@code ContentProvider}, or {@code BroadcastReceiver}. In these cases, this lifecycle reports
+ * itself in the {@link Lifecycle.State#CREATED} state.
+ * <p>
+ * Regardless of what happens to a backing {@code Activity} or @{code Service}, this lifecycle
+ * will only report itself as {@link Lifecycle.State#DESTROYED} when the associated {@link FlutterEngine}
+ * itself is destroyed. This is because a {@link Lifecycle} is not allowed to emit any events after
+ * going to the {@link Lifecycle.State#DESTROYED} state. Thus, this lifecycle cannot emit such an
+ * event until its associated {@link FlutterEngine} is destroyed. This then begs the question, what
+ * happens when the backing {@code Activity} or {@code Service} is destroyed? This lifecycle will
+ * report the process up to the {@link Lifecycle.Event#ON_STOP} event, but will ignore the
+ * {@link Lifecycle.Event#ON_DESTROY} event. At that point, this lifecycle will be back in its
+ * default {@link Lifecycle.State#CREATED} state until some other backing {@link Lifecycle} is
+ * registered.
+ */
+class FlutterEngineAndroidLifecycle extends LifecycleRegistry {
```

Comment:
nit: This class should probably be final. If it's package private we're not expecting it to be extended.

Reply:
Done.